src/commands/plugins.rs

```diff
@@ -0,0 +1,275 @@
+
+    /// Path to remote plugin manifest.
+    #[clap(
+        name = "REMOTE_PLUGIN_MANIFEST",
+        short = 'u',
+        long = "url",
+        conflicts_with = "LOCAL_PLUGIN_MANIFEST"
+    )]
+    pub remote_manifest_src: Option<Url>,
+
+    /// Skips prompt to accept the installation of the plugin[s].
+    #[clap(short = 'y', long = "yes", takes_value = false)]
+    pub yes_to_all: bool,
+
+    /// Specific version of a plugin to be install from the centralized plugins
+    /// repository.
+    #[clap(
+        long = "version",
+        short = 'v',
+        conflicts_with = "REMOTE_PLUGIN_MANIFEST",
+        conflicts_with = "LOCAL_PLUGIN_MANIFEST",
+        conflicts_with = "ALL",
+        requires("PLUGIN_NAME")
+    )]
+    pub version: Option<Version>,
+
+    /// Allow downgrading a plugin's version.
+    #[clap(short = 'd', long = "downgrade", takes_value = false)]
+    pub downgrade: bool,
+}
+
+impl Upgrade {
+    /// Upgrades one or all plugins by reinstalling the latest or a specified
+    /// version of a plugin. If downgrade is specified, first uninstalls the
+    /// plugin.
+    pub async fn run(self) -> Result<()> {
+        let plugins_dir = get_spin_plugins_directory()?;
+        let spin_version = env!("VERGEN_BUILD_SEMVER");
+        let manifest_dir = plugins_dir.join(PLUGIN_MANIFESTS_DIRECTORY_NAME);
+
+        // Check if no plugins are currently installed
+        if !manifest_dir.exists() {
+            println!("No currently installed plugins to update.");
+            return Ok(());
+        }
+
+        if self.all {
+            // Install the latest of all currently installed plugins
+            for plugin in std::fs::read_dir(manifest_dir)? {
+                let path = plugin?.path();
+                let name = path
+                    .file_stem()
+                    .ok_or_else(|| anyhow!("expected directory for plugin"))?
+                    .to_str()
+                    .ok_or_else(|| anyhow!("Cannot convert directory to String"))?
+                    .to_string();
+                if let Err(e) = PluginInstaller::new(
+                    ManifestLocation::PluginsRepository(PluginInfo::new(
+                        &name,
+                        Url::parse(SPIN_PLUGINS_REPO)?,
+                        None,
+                    )),
+                    plugins_dir.clone(),
+                    self.yes_to_all,
+                    spin_version,
+                )
+                .install()
+                .await
+                {
+                    // Ignore plugins that were not installed from the central
+                    // plugins repository
+                    if e.to_string().contains("Could not find plugin") {
+                        log::info!(
+                            "Could not update {} plugin as DNE in central repository",
+                            name
+                        );
+                    } else {
+                        return Err(e);
```

Comment:
Correct to bail on error rather than upgrading remaining plugins?  This could mean one bad plugin blocking others from upgrading.

Reply:
made it so upgrading continues. Good call out!